src/server/scripts/EasternKingdoms/ShadowfangKeep/shadowfang_keep.cpp

```diff
@@ -205,6 +205,131 @@ class npc_arugal_voidwalker : public CreatureScript
+};
+
+enum ArugalTexts
+{
+    SAY_ARUGAL_AGGRO            = 1, // You, too, shall serve!
+    SAY_ARUGAL_TRANSFORM        = 2, // Release your rage!
+    SAY_ARUGAL_KILLED_UNIT      = 3  // Another falls!
+};
+
+enum ArugalEvents
+{
+    EVENT_ARUGAL_VOID_BOLT      = 1,
+    EVENT_ARUGAL_TELEPORT       = 2,
+    EVENT_ARUGAL_THUNDERSHOCK   = 3,
+    EVENT_ARUGAL_CURSE          = 4
+};
+
+std::array<uint32, 3> arugalTeleportSpells =
+{
+    SPELL_TELE_UPPER,
+    SPELL_TELE_STAIRS,
+    SPELL_TELE_SPAWN
+};
+
+class boss_archmage_arugal : public CreatureScript
+{
+public:
+    boss_archmage_arugal() : CreatureScript("boss_archmage_arugal") { }
+
+    struct boss_archmage_arugalAI : public BossAI
+    {
+        boss_archmage_arugalAI(Creature* creature) : BossAI(creature, BOSS_ARUGAL) { }
+
+        void Reset() override
+        {
+            _Reset();
+        }
+
+        void KilledUnit(Unit* who) override
+        {
+            if (who->GetTypeId() == TYPEID_PLAYER)
+                Talk(SAY_ARUGAL_KILLED_UNIT);
+        }
+
+        void SpellHitTarget(Unit* /*target*/, SpellInfo const* spell) override
+        {
+            if (spell->Id == SPELL_ARUGAL_CURSE)
+                Talk(SAY_ARUGAL_TRANSFORM);
+        }
+
+        void EnterCombat(Unit* /*who*/) override
+        {
+            _EnterCombat();
+            Talk(SAY_ARUGAL_AGGRO);
+            events.ScheduleEvent(EVENT_ARUGAL_CURSE, Seconds(7));
+            events.ScheduleEvent(EVENT_ARUGAL_TELEPORT, Seconds(15));
+            events.ScheduleEvent(EVENT_ARUGAL_VOID_BOLT, Seconds(1));
+            events.ScheduleEvent(EVENT_ARUGAL_THUNDERSHOCK, Seconds(10));
+        }
+
+        void AttackStart(Unit* who) override
+        {
+            if (!who)
+                return;
+
+            AttackStartCaster(who, 100.0f); // void bolt range is 100.f
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (!UpdateVictim())
+                return;
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            events.Update(diff);
+
+            while (auto eventId = events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                    case EVENT_ARUGAL_CURSE:
+                        if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 1, 30.0f, true))
+                            DoCast(target, SPELL_ARUGAL_CURSE);
+
+                        events.ScheduleEvent(eventId, Seconds(15));
+                        break;
+                    case EVENT_ARUGAL_TELEPORT:
+                    {
+                        auto spell = arugalTeleportSpells.front();
+                        DoCast(spell); 
+                        // rotate array, can be done with some kind of shuffle. Up to you to figure out what is best.
+                        std::rotate(arugalTeleportSpells.begin(), arugalTeleportSpells.begin() + 1, arugalTeleportSpells.end());
```

Comment:
you CANNOT modify a global variable in EVERY class instance of boss_archmage_arugalAI without any locking mechanism, it will cause the very much dreaded race conditions. I strongly suggest you to spend some time reading up about "race condition" to fully understand what and why is bad.